src/BusinessLayer/TelemetryReporting.cpp

```diff
@@ -1,215 +1,465 @@
 {
-   const unsigned int unframedPacketLength = BATTERY_DATA_LENGTH + CHECKSUM_LENGTH;
-   unsigned char packetPayload[unframedPacketLength];
-   packetPayload[0] = BATTERY_DATA_ID;
-   writeFloatIntoArray(packetPayload, 1, vehicleData_.batteryVoltage);
-   writeFloatIntoArray(packetPayload, 5, vehicleData_.batteryCurrent);
-   writeFloatIntoArray(packetPayload, 9, vehicleData_.packStateOfChargePercentage);
-   writeFloatIntoArray(packetPayload, 13, vehicleData_.balancePackStateOfCharge);
-   packetPayload[17] = static_cast<unsigned char>(vehicleData_.secondaryBatteryUnderVoltage);
-
-   addChecksum(packetPayload, BATTERY_DATA_LENGTH);
-   unsigned char packet[unframedPacketLength + FRAMING_LENGTH_INCREASE];
-   unsigned int packetLength = frameData(packetPayload, unframedPacketLength, packet);
-   communicationService_.sendData(packet, packetLength);
+    qDebug() << "Sending Battery";
+    const unsigned int unframedPacketLength = BATTERY_LENGTH + CHECKSUM_LENGTH;
+    unsigned char packetPayload[unframedPacketLength];
+
+    packetPayload[0] = CcsDefines::BATTERY_PKG_ID;
+    bool aliveArray[] = {batteryData_.alive};
+    writeBoolsIntoArray(packetPayload, 1, aliveArray, 1);
+    writeFloatIntoArray(packetPayload, 2, batteryData_.packSocAmpHours);
+    writeFloatIntoArray(packetPayload, 6, batteryData_.packSocPercentage);
+    writeFloatIntoArray(packetPayload, 10, batteryData_.packBalanceSoc);
+    writeFloatIntoArray(packetPayload, 14, batteryData_.packBalanceSocPercentage);
+    writeUShortIntoArray(packetPayload, 18, batteryData_.chargingCellVoltageError);
+    writeUShortIntoArray(packetPayload, 20, batteryData_.cellTemperatureMargin);
+    writeUShortIntoArray(packetPayload, 22, batteryData_.dischargingCellVoltageError);
+    writeUShortIntoArray(packetPayload, 24, batteryData_.totalPackCapacity);
+    bool contactorArray[] = {batteryData_.contactor0Errorstatus,
+                             batteryData_.contactor1ErrorStatus,
+                             batteryData_.contactor0Status,
+                             batteryData_.contactor1Status,
+                             batteryData_.contactor12VSupplyOk,
+                             batteryData_.contactor2ErrorStatus,
+                             batteryData_.contactor2Status
+                            };
+    writeBoolsIntoArray(packetPayload, 26, contactorArray, 7);
+    packetPayload[27] = (unsigned char)batteryData_.prechargeState;
+    bool prechargeArray[] = {batteryData_.prechargeTimerElapsed,
+                             batteryData_.prechargeTimerNotElapsed
+                            };
+    writeBoolsIntoArray(packetPayload, 28, prechargeArray, 2);
+    writeUShortIntoArray(packetPayload, 29, batteryData_.prechargeTimerCount);
+    writeUShortIntoArray(packetPayload, 31, batteryData_.lowestCellVoltage);
+    packetPayload[33] = batteryData_.lowestCellVoltageCmuNumber;
+    packetPayload[33] += batteryData_.lowestCellVoltageCellNumber << 2;

```

Comment:
So you have your field of 1 char (0b0000 0000). The CMU number goes in 0b0000 XXXX and the cell number goes 0bXXXX 0000. To get the CMU number in you need to select the lowest 4 bits and assign them. In this case it is fine to just do the straight assignment like you did.

Now for the cell number it currently looks like this 0b0000 XXXX and you want to put it in the field spot 0bXXXX 0000. To do so you have to shift it over 4 bits ( also called 1 nibble).  Once you do that you now have to OR it into the field (not the same as the current addition). So to sum it up the code would be:
```
field = CMU & 0xFF; // the masking is optional but safer
field |= CELL << 4;
```